ospf6d/ospf6_intra.c

```diff
@@ -623,52 +574,26 @@ void ospf6_network_lsa_originate(struct event *thread)
 static char *ospf6_link_lsa_get_prefix_str(struct ospf6_lsa *lsa, char *buf,
 					   int buflen, int pos)
 {
-	char *start, *end, *current;
-	struct ospf6_link_lsa *link_lsa;
-	struct in6_addr in6;
-	struct ospf6_prefix *prefix;
-	int cnt = 0, prefixnum;
+	struct ospf6_link_lsa *link_lsa = lsa_after_header(lsa->header);
+	struct ospf6_prefix *prefix = nth_prefix(lsa->header, pos);
+	struct in6_addr in6 = { 0 };
 
-	if (lsa) {
-		link_lsa = (struct ospf6_link_lsa
-				    *)((caddr_t)lsa->header
-				       + sizeof(struct ospf6_lsa_header));
+	if (!lsa || !prefix || !buf || buflen < (1 + INET6_ADDRSTRLEN))
+		return NULL;
 
-		if (pos == 0) {
-			inet_ntop(AF_INET6, &link_lsa->linklocal_addr, buf,
-				  buflen);
-			return (buf);
-		}
+	/* position zero is used for the lladdr in the body of the LSA */
+	if (pos == 0) {
+		inet_ntop(AF_INET6, &link_lsa->linklocal_addr, buf, buflen);
+		return buf;
+	}
 
-		prefixnum = ntohl(link_lsa->prefix_num);
-		if (pos > prefixnum)
-			return NULL;
+	memcpy(&in6, OSPF6_PREFIX_BODY(prefix),
```

Comment:
Should we consider a function to do this copy?

Reply:
Sorry, I'm not sure I fully understand the intent. Are you suggesting a function wrapping memcpy, in order to add explicit types, to improve the issue of potentially providing an incorrect addresses or length? It could be a number of different one-line function variations for different source and destination types. I will try it and see what it looks like.

Reply:
I just see us using the same logic in two places ... and I can where we might want to use this other places and/or provide a check at some point ... ?? Not a blocker, just asking if we should consider it.

Reply:
I can see what you mean now. There is such a function, called `ospf6_prefix_in6_addr`, so it is definitely an opportunity to improve consistency.

The implementation of that function seemed a bit unintuitive to me, and I see it was added in commit b8ce0c36962 to address an OOB read flagged by Coverity, so I should take the time to understand it better.